Last implementation of package classification was very slow. It needed

guliashvili opened the pull request:
linear time in terms of number of different possible packages. My
implementation does preprocessing. It creates array and uses key as
index. However each array can be leaf or can need another array. (Some
packaged need 3 key or 2 key)

guliashvili commented:
Hi Guys, As you see it's my first commit here. It's pretty nice project I'd love to optimize many stuff you help me of course :)

guliashvili commented:
anyone? knock knock :( :-1:

Zealar commented:
Hello can post some test results? Also please use l2j formatter for code.

guliashvili commented:
Hi,
What kind of test results?
Where can I find  l2j formatter for code?

Zealar commented:
You say "Last implementation of package classification was very slow." so what is real difference? 1ms 2ms? What we win if change it to that way.
Is inside https://github.com/L2J/L2J_Server/tree/develop/.settings
is auto format code when use eclipse.

guliashvili commented:
There is about 500 package, finding random package using last implementation needed checking 250 package on average. With my implementation it's done with cost of 2-5 package check.

Zoey76 commented on src/main/java/com/l2jserver/gameserver/network/fastforwarder/FastForwarderTable.java:
```diff
@@ -0,0 +1,65 @@
+
+public abstract class FastForwarderTable implements FastForwarder
+{
+	private static final Logger _log = Logger.getLogger(FastForwarderTable.class.getName());
```
Logger should use SLF4J.

Zoey76 commented:
Do you want to use this to replace which code exactly?

L2GamePacketHandler#handlePacket(ByteBuffer buf, L2GameClient client) ?

guliashvili commented:
@Zoey76 
Before, whole checking of what packet we got was in 
	[public ReceivablePacket<L2GameClient> handlePacket(ByteBuffer buf, L2GameClient client)](https://github.com/L2J/L2J_Server/blob/develop/src/main/java/com/l2jserver/gameserver/network/L2GamePacketHandler.java)

Now,
I do linear preprocessing and build static table, table is kind of hashmap, but because keys are small I can even do faster thing and use poor arrays without hashes and use keys directly as an array indexes.
So, new [handlePacket function](https://github.com/guliashvili/L2J_Server/blob/develop/src/main/java/com/l2jserver/gameserver/network/L2GamePacketHandler.java#L649) is just one line. Which uses this pre created forwarding table.
Basically I changed everything in L2GamePacketHandler and also added some helper classes.

Zoey76 commented:
If you want to micro-optimize the packet selection, you should read, if you haven't already, about the under-the-hood implementations of switch statement in Java.

[3.10. Compiling Switches ](http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-3.html)

Java uses two modes tableswitch, with Big-O of O(1), corresponding to your implementation and lookupswitch which uses a sorted pair table, with Big-O of O(log(n)), which for 500 packets would be 2.6~ top operations.

How do you plan to create the List<FastForwarderPair> list ?

guliashvili commented:
Good job java optimizations. :) .
I will check how exactly it is optimized and if my optimization is faster

Sacrifice78 commented:
Any news or improvement about that or any comment at respect???

UnAfraid commented:
That wouldn't increase performance a lot unless there are 2000-3000 players, even then you might not notice much of a difference
